### PR TITLE
Add observation for message channels

### PR DESCRIPTION
**Pull request by artembilan**

### CODE REVIEW

**Review by garyrussell — CHANGES_REQUESTED** [review]
Commit comment would have been helpful.

File: spring-integration-core/src/main/java/org/springframework/integration/channel/AbstractMessageChannel.java [hunk 1]
```diff
@@ -293,11 +312,45 @@ public boolean send(Message<?> messageArg, long timeout) {
...
+		else if (this.metricsCaptor != null) {
+			return sendWithMetrics(message, timeout);
+		}
+		else {
+			return sendInternal(message, timeout);
+		}
+	}
+
+	private boolean sendWithObservation(Message<?> message, long timeout) {
+		MutableMessage<?> messageToSend = MutableMessage.of(message);
```
[the remark this answers is not in the document]
Comment (artembilan) [reply spring-integration-core/src/main/java/org/springframework/integration/channel/AbstractMessageChannel.java:327]: See the `MessageSenderContext` implementation.
The `MessageChannel.send()` is really a point in Spring Integration where we produce messages.
According to `PRODUCER` entity of the observation, we have to expose some `Setter` which would carry on values from the current context to downstream consumers - propagation, essentially.
Therefore a `MutableMessage` to be able to modify headers in the trace `Propagator`.
See `ObservationPropagationChannelInterceptorTests`.

Or is your "Why?" about this new `of()` factory method? 😄 

Sure! I can add more info into commit message if we are really on the same page about the solution after this review and discussion.

Thank you for understanding!

P.S. Perhaps I deliberately have left so little info in the PR, therefore you would pay attention for what is going on and we would have a healthy discussion like this 😉

[the remark this answers is not in the document]
Comment (garyrussell) [reply spring-integration-core/src/main/java/org/springframework/integration/channel/AbstractMessageChannel.java:327]: Yes, I missed the `getHeaders().put(...)`; now I see it.

I will be honest in that I still don't see why we need to observe channel sends, it seems much too fine-grained to me.

[the remark this answers is not in the document]
Comment (artembilan) [reply spring-integration-core/src/main/java/org/springframework/integration/channel/AbstractMessageChannel.java:327]: Well, if channel is distributed, then we just lose the producer point of the trace.
It is really up to end-user now what to chose for instrumentation. 
Perhaps they indeed would apply only inbound and outbound channel adapters.
However the current solution with Sleuth is only channel instrumentation 🤷

[the remark this answers is not in the document]
Comment (garyrussell) [reply spring-integration-core/src/main/java/org/springframework/integration/channel/AbstractMessageChannel.java:327]: Ok; just address my other comment and will merge.

**Review by garyrussell — CHANGES_REQUESTED** [review]
Doc polishing.

**Pull request merged**